webapp/src/ts/services/select2-search.service.ts

```diff
@@ -77,17 +83,39 @@ export class Select2SearchService {
     this.searchService
       .search('contacts', filters, searchOptions)
       .then((documents) => {
-        if (currentQuery === params.data.q) {
-          successCb({
-            results: options.sendMessageExtras(this.prepareRows(documents)),
-            pagination: {
-              more: documents.length === options.pageSize
-            }
-          });
+        if (this.currentQuery !== search) {
+          return;
         }
+
+        if (this.onContactSelect) {
+          this.selectEl!.off('select2:select', this.onContactSelect);
+          this.onContactSelect = null;
+        }
+
+        this.onContactSelect = async (event) => {
+          this.onContactSelect = null;
+
+          const docId = event.params?.data?.id;
+          if (!search || !docId) {
+            return;
+          }
+
+          const doc = await this.getDoc(docId);
+          await this.searchTelemetryService.recordContactByTypeSearch(doc, search);
```

Comment:
Minor, but technically I think this will only hit the `contacts_by_type_freetext` view if the `types` param above is valued.  Otherwise, I guess `shared-libs/search` will just route this to `contacts_by_freetext`...

Reply:
Right, but `select2-search` is only ever used to retrieve contacts by type, no?

Reply:
Ah no you're right, there is one case where this service is used without type `types` parameter
https://github.com/medic/cht-core/blob/b7b2629bfe92329f9974e94adc6786bc687f3202/webapp/src/ts/modals/send-message/send-message.component.ts#L187

I'll handle this scenario too, thanks!

Reply:
`Select2SearchService` is initialized twice in the codebase. First occurrence in https://github.com/medic/cht-core/blob/b7b2629bfe92329f9974e94adc6786bc687f3202/webapp/src/ts/modals/send-message/send-message.component.ts#L174-L189
And the second occurrence in https://github.com/medic/cht-core/blob/59b42e2fa3271a69c713b4720b9ee9fb7ea3dc61/webapp/src/ts/modals/edit-report/edit-report.component.ts#L42-L53

Both times it's initialized with the `types` parameter and then that gets translated into a query against `medic-client/contacts_by_type` in https://github.com/medic/cht-core/blob/843888c13456919f50daf12449611ed28c732e5c/shared-libs/search/src/generate-search-requests.js#L137-L148
and its results get "merged" with the freetext query results in https://github.com/medic/cht-core/blob/843888c13456919f50daf12449611ed28c732e5c/shared-libs/search/src/generate-search-requests.js#L274-L276

Reply:
Select2Search is also inited by the db-object-widget: https://github.com/medic/cht-core/blob/45478853727d40b60b78c6359fb11fc7606d4a9c/webapp/src/js/enketo/widgets/db-object-widget.js#L59

I did some double-checking and confirmed that when you use the `select-contact` appearance _without a `type-*` appearance, it will call through to just the `contacts_by_freetext` view (since the types in the search request are empty).

Reply:
🫡 your enketo-fu is always appreciated, I'll handle this case

Reply:
Okay I pushed what I think will cover this scenario but I'm about to drop off. I've noted to myself to manually test it and also add unit tests for this particular case

Reply:
Okay there was no straightforward way to unit test this so I went for an e2e test :slightly_smiling_face:

Reply:
:+1: You would think in 2024 we would have a half-way decent way to unit-test code that involves jquery managed browser events.  So painful....